### ISSUE TITLE
Some questions about code and paper

### ISSUE THREAD

**Issue by xiaomingjie** [jackd/template_ffd#5]
Hi @jackd ,

These days I've looked into your work and got some questions.

1. I never used `tf.estimator` before and there is few open source projects using it. It seems that `TemplateFfdBulder.initilalize_variables` calls `get_estimator_spec` with input [features=(one_example_id,image),labels=point_cloud]. Inside it, we get output of the network and define loss. But I think there is only one network and one loss. What if I would like to train two sub_network such as GAN, how can I define loss and train?

2. According to your paper, you learn the offset of control points from the network. I think you got your result in `TemplateFfdBulder.get_inference` with output dict(example_id=example_id, probs=probs, dp=dp). And then I think you used probs to get entropy loss and dp to get inferred point clouds, but I don't know how and where you choose the final template to be deformed.

3. keep looking into `get_dataset` but still stuck with dataset, where do you put it into network batch by batch?

Looking forward to your reply and thanks!

**Comment by jackd** [comment]
1. `get_estimator_spec` is a `model_fn` which can be passed to `tf.estimator.Estimator`'s constructor. That said, it's not the -only- use. For example, in `initialize_variables` I use it to save variables saved in a `tf.keras` model to 'normal' tensorflow checkpoint format. It's a dirty hack that is required to make `keras` play nicely with `Estimator`'s.

2. Tensorflow and `Estimator`s are responsible for training the network and generating inferences (`probs` and `dps`). After that, all manipulations are done using standard numpy, with `dids` datasets handling the saving of each stage.

3. `Estimator`s handle all session management. For example, `ModelBuilder.train` is a thin wrapper around `Estimator.train`. They handle variable loading/saving, graph/summary writing and a lot of boilerplate that you don't want to have to write for every single model. There are certainly instances where they aren't appropriate, in which case falling back to lower level ops is required, but I find they can be made to fit a large majority of cases. For example, if you have multiple `train_op` like in a GAN, you can simply create an operation that combines the two.

```
with tf.control_dependencies([generator_train_op, critic_train_op]):
    train_op = tf.no_op()
...
return tf.estimator.EstimatorSpec(train_op=train_op)
```

That said, if you're interested in GANs you should check out `tf.contrib.gan.GANEstimator`. It's a tiny bit more restrictive than I'd like, but its super convenient for proto-typing GANs with relatively standard architectures.

Again, just to stress, Estimator's are not the be-all and end-all of tensorflow. You don't -have- to use them, but if you do you aren't restricted to ONLY using them. You can train a model with an Estimator and use the resulting weights in a standard graph with complex Estimator-incompatible functionality. Don't be afraid to walk away from them if your use case doesn't benefit. They're one tool in a workshop. For what they do, I think they're amazing, but if you try and use a hammer to put in a screw you're going to be unhappy with the hammer's performance.

**Comment by xiaomingjie** [comment]
Now the question comes to input data..
`template_ffd_builder.get_image_features` leads to `_get_mobilenet_features`. The image turns into `input_shape=image.shape.as_list()[1:]`. Here what shape is the `input_shape`? Is it [32,192,256,3]? This really confuses me when I want to input data to other network..

**Comment by jackd** [comment]
Its the input shape of the image you passed in, skipping the batch dimensions. My images are all [192 * 256], so `input_shape==[192, 256, 3]`.

**Comment by xiaomingjie** [comment]
So when and where load data batch by batch in training?
I think this is still a data manipulation question..

**Comment by jackd** [comment]
I pass that off to the estimator to handle.

**Comment by xiaomingjie** [comment]
There it is!
I got it. It means if I use estimator I don't need to care about how data be loaded in network. If I don't use estimator, I should use dataset in another way for example new a thread to load data batch by batch as we do normally. Am I right?

**Comment by jackd** [comment]
... no. You still need to tell the estimator how to load the data, you just don't call the function directly. There are a number of ways to load data. I use tf.fata.Datasets. see builder.get_inputs

**Comment by xiaomingjie** [comment]
HI @jackd ,
I just used your dataset in PSGN and found some issues.

1. I used `get_train_inputs()` to get 'plane' data and put it into PSGN. But after 50 iterations, memory usage achieved 95% and my PC nearly crashed... I suppose I used `get_train_inputs()` incorrectly just like I load all data once instead of batch by batch which causes memory overflow. So how can I avoid it?
```
builder = get_builder('plane')
feature,pointcloud = builder.get_train_inputs()
image = feature['image']
data,ptcloud=sess.run([image,pointcloud])  #if I don't do this, error occured 'tpye cannot be tensor'
_,pred,total_loss,trainloss,trainloss1,trainloss2,distmap_0=sess.run([optimizer,x,loss,loss_nodecay,dists_forward,dists_backward,dist0],feed_dict={img_inp:data,pt_gt:ptcloud,training_flag:True})
```

2. I also found that `get_train_inputs` returns ground truth point cloud with `n_resample=1024`. It seems that GT has only 1024 points not 16384 at training but I think it should be 16384...

3. It seems that train loss got very low at the very beginning just like 0.02 or 0.007 after iteration 30/40, while PSGN's original dataset achieved 0.02 after 1000 iteration. I think it's weird...

hope for your reply and hope that I'm not bordering you....

**Comment by jackd** [comment]
1. Use feed dicts or the output of `get_train_inputs` - not both. i.e. wherever you defined `img_inp` and `pt_gt` as placeholders, just get them from the result of `get_train_inputs`.
2. I use 16384 as the number of points in an initial densely sampled cloud, then resample 1024 points for training. This gives a slightly less varied cloud than if we resample the original mesh each time, but is computationally much more efficient. You should be able to change the number of resampled points using `params`.
3. Since we start from minimally deformed templates, this doesn't surprise me, particularly for planes since most planes have the same basic structure.

**Comment by xiaomingjie** [comment]
1. I defined placeholder in building graph, so should I just use img_inp = feature['image'] right after img_inp = tf.placeholder()? Then how can I input the real input when start sess.run()?

2. To the question 3. But I only use your data (RGB and GTpointcloud) instead of the original data of PSGN and put it into PSGN without any other changes. So output is point cloud not ffd parameters becauce I didn't use ffd and deformed templates. In this situation, the chamfer loss performs well at the first beginning..

**Comment by jackd** [comment]
1. If you want to use your data, use your data with a palceholder. If you want to use my data, use my data and no placeholder.
2. I'm not familiar with PSGN code. From their paper I believe they trained a single model across all categories (plus more... something like 220k models if I remember correctly) rather than a seperate model for each category, which would affect initial training speed, though assuming the image processing network has sufficient capacity probably won't affect the long-term accuracy much.

**Comment by xiaomingjie** [comment]
I think I got you. Just use image = feature['image'] and then put image into network directly in building graph phase. I will try it. 
Thanks very much!
